scheduler/reconcile_util.go

```diff
@@ -216,15 +216,28 @@ func (a allocSet) fromKeys(keys ...[]string) allocSet {
 // 4. Those that are on nodes that are disconnected, but have not had their ClientState set to unknown
 // 5. Those that are on a node that has reconnected.
 // 6. Those that are in a state that results in a noop.
-func (a allocSet) filterByTainted(taintedNodes map[string]*structs.Node, supportsDisconnectedClients bool, now time.Time) (untainted, migrate, lost, disconnecting, reconnecting, ignore allocSet) {
+func (a allocSet) filterByTainted(taintedNodes map[string]*structs.Node, serverSupportsDisconnectedClients bool, now time.Time) (untainted, migrate, lost, disconnecting, reconnecting, ignore allocSet) {
 	untainted = make(map[string]*structs.Allocation)
 	migrate = make(map[string]*structs.Allocation)
 	lost = make(map[string]*structs.Allocation)
 	disconnecting = make(map[string]*structs.Allocation)
 	reconnecting = make(map[string]*structs.Allocation)
 	ignore = make(map[string]*structs.Allocation)
 
+	supportsDisconnectedClients := serverSupportsDisconnectedClients
+
 	for _, alloc := range a {
+
+		// make sure we don't apply any reconnect logic to task groups
+		// without max_client_disconnect
+		if alloc.Job != nil {
+			tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+			if tg != nil {
+				supportsDisconnectedClients = serverSupportsDisconnectedClients &&
+					tg.MaxClientDisconnect != nil
+			}
+		}
```

Comment:
Adding this didn't cause any `scheduler` unit tests to fail when I ran them locally, so I imagine I'm going to want to come back thru and add some test cases before we merge this. 😀

Reply:
Alternate for this approach: should we gate _all_ the logic of this function behind `supportsDisconnectedClients` and then have the old (1.2.6) logic here too? Or is that just way too gross?

Reply:
It's not gross per se. I thought about 2 functions at the beginning. And in my canary fix branch, I brought that function back as `filterCanaries`, but it didn't work as-is, because there is the disconnected client aspect to deal with. It's worth a spike and might make this code smell less 😺

Reply:
I don't think we can do 2 clean functions, because we might have an allocSet where some of the allocs have `max_client_disconnect` and some don't (I think?)

Reply:
I was thinking maybe we could just say... `if !serverSupportsDisconnect { callV1 }` in essence. But yeah, it's challenging all around.  Also, I've got updates in flight related to canaries. Let me know if you want to join forces. Anyway that we can make this require less constant context awareness would be great.

Reply:
I've just pushed up a change to expand test coverage (I did some coverage profile-guided tests to get us a bit closer to complete coverage for this function) but I've avoided doing that refactoring given that you've got work in-flight. That'll keep this PR small, let you land your fix, and then we can come back to do a refactor PR as a mop-up.